Moves <TextHighlight /> component to @wordpress/components package

getdave opened the pull request:
Closes https://github.com/WordPress/gutenberg/issues/18074.

This PR extracts the `TextHighlight` used within `LinkControl` into a standalone component.

`TextHighlight` is a component which applies a `<mark>` tag around all/any matching strings within a string of text. This is useful for highlighting search terms within a string of text. 

## How has this been tested?
* Manually verifying `LinkControl` hasn't broken and search terms are still highlighted.
* Checking Storybook behaviour.

## Screenshots <!-- if applicable -->

## Types of changes
New feature (non-breaking change which adds functionality).

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->.

getdave committed:
Moves component to @wordpress/components package

Also updates LinkControl with new reference

getdave committed:
Write a little Story…

getdave committed:
Write test coverage

getdave committed:
Add docs

getdave committed:
Auto docs manifest update

gziolo commented on packages/components/src/text-highlight/README.md:
```diff
@@ -0,0 +1,39 @@
+
+const MyTextHighlight = () => (
+	<TextHighlight
+		text={ 'Why do we like Gutenberg? Because Gutenberg is the best!' }
```
Nit:
```suggestion
		text="Why do we like Gutenberg? Because Gutenberg is the best!"
```

gziolo commented on packages/components/src/text-highlight/README.md:
```diff
@@ -0,0 +1,39 @@
+const MyTextHighlight = () => (
+	<TextHighlight
+		text={ 'Why do we like Gutenberg? Because Gutenberg is the best!' }
+		highlight={ 'Gutenberg' }
```
Nit:
```suggestion
		highlight="Gutenberg"
```

gziolo commented on packages/components/src/text-highlight/stories/index.js:
```diff
@@ -0,0 +1,22 @@
+export default { title: 'Components|TextHighlight', component: TextHighlight };
+
+export const _default = () => {
+	const defaultText = 'We call the new editor Gutenberg. The entire editing experience has been rebuilt for media rich pages and posts.';
```
Nit: this const can be inlined.

getdave replied:
Fixed.

gziolo reviewed (commented):
I was wondering whether this component could be refactored to use this new experimental API just introduced by @nerrad in #17376:
`__experimentalCreateInterpolateElement`
#18623 illustrates how it can be used in action. In this case, it would be possible to avoid wrapping text with `span` elements.

getdave commented:
Thanks I'm going to get back to this just as soon as I can. Likely this Friday.

getdave committed:
Update packages/components/src/text-highlight/README.md

Co-Authored-By: Grzegorz (Greg) Ziółkowski <grzegorz@gziolo.pl>

getdave committed:
Update packages/components/src/text-highlight/README.md

Co-Authored-By: Grzegorz (Greg) Ziółkowski <grzegorz@gziolo.pl>

getdave committed:
Fix inlining of var

Addresses https://github.com/WordPress/gutenberg/pull/18609#discussion_r348392899

getdave committed:
Updates to utilise new __experimentalCreateInterpolateElement API

Addresses https://github.com/WordPress/gutenberg/pull/18609#pullrequestreview-319722852

getdave commented:
> I was wondering whether this component could be refactored to use this new experimental API just introduced by @nerrad in #17376:
> `__experimentalCreateInterpolateElement`
> #18623 illustrates how it can be used in action. In this case, it would be possible to avoid wrapping text with `span` elements.

@gziolo I was intending to keep the `<span>` as it might be useful to be able to style things but I've realised that perhaps it's not necessary as CSS will let us style the default case and have overides for the highlight text anyway. 

As a result, I've refactored to the new experimental API.

getdave committed:
Remove unnecessary span

gziolo commented on packages/components/src/text-highlight/index.js:
```diff
@@ -16,13 +16,17 @@ const TextHighlight = ( { text = '', highlight = '' } ) => {
-				regex.test( part ) ? <mark key={ i }>{ part }</mark> : <span key={ i }>{ part }</span>
-			) ) }
-		</Fragment>
+	const parts = text.split( regex ).filter( ( part ) => part );
```
Could it use [String.replace](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace) with the RegExp or do I miss something specific in the logic?

getdave replied:
Actually I think it's fine. Probably me overengineering things.

getdave committed:
Updates to simplify implementation to String.replace

gziolo reviewed (approved):
It looks like you need to regenerate one of the snapshots generated from Storybook story. Otherwise, Jest seems to be happy about the proposed refactor :)

Nice work, thanks for applying my feedback.

Feel free to merge once you get into agreement with Travis 😄

getdave committed:
Update snapshots

getdave commented:
Thanks for the reviews @gziolo! Much appreciated.
